### ISSUE TITLE
Better error message for overloading functions that return union type

### ISSUE THREAD

**Issue by binh-vu** [microsoft/pylance-release#2020]
Networkx, a popular library for working with graph, uses lots of overloading. One example is iterating through edges in the graph. It can returns an iterator of tuples of 3 or 4 items (if `keys` argument is true). For example:

```
g = nx.MultiDiGraph()
...
# pylance complains that: Tuple size mismatch: expected 3 but received 4
for uid, vid, edge_data in g.edges(data=True):
    pass

# pylance compalins that: Tuple size mismatch: expected 4 but received 3
for uid, vid, eid, edge_data in g.edges(data=True, keys=True):
    pass
```

The error message is quite confusing because at first it told me to use a tuple of 4, but then when I use tuple of 4, it told me to use a tuple of 3. It would be better if the error message could mention the union type so the developers know that they need to do type narrowing or use type guard.

**Comment by erictraut** [comment]
Thanks for reporting this. I agree the current behavior is confusing.

**Comment by erictraut** [comment]
This will be addressed in the next release of pylance.

**Comment by erictraut** [comment]
The new diagnostic message is a multi-line message with additional details. This is consistent with other diagnostic messages provided by pylance.

![Screen Shot 2021-11-02 at 1 21 10 AM](https://user-images.githubusercontent.com/7040122/139810612-5b646e96-081b-44b5-b5c6-fb4a54e9dcd3.png)

**Comment by heejaechang** [comment]
fixed in 2021.11.0
